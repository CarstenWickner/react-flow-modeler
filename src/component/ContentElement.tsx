import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";

export class ContentElement extends React.Component<{
    elementId: string;
    editMenu: (() => React.ReactNode) | undefined;
    onSelect: (elementId: string) => void;
}> {
    onClick = (event: React.MouseEvent): void => {
        const { onSelect, elementId } = this.props;
        onSelect(elementId);
        event.stopPropagation();
    };

    render(): React.ReactNode {
        const { editMenu, children } = this.props;
        return (
            <>
                <div className="stroke-horizontal arrow" />
                <div className="flow-element-wrapper">
                    <div className={`flow-element content-element${editMenu ? " selected" : ""}`} onClick={this.onClick}>
                        {children}
                    </div>
                    {editMenu && editMenu()}
                </div>
                <HorizontalStroke optional />
            </>
        );
    }
}
