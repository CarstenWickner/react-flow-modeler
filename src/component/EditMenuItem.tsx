import * as React from "react";
import { useDrag } from "react-dnd";

import { DraggedLinkContext } from "../types/EditAction";

const disabledDragging: [{ isDragging: boolean }, React.LegacyRef<HTMLSpanElement>] = [{ isDragging: false }, undefined];

export const EditMenuItem: React.FC<{
    options: { className?: string; title?: string };
    defaultClassName: string;
    onClick?: (event: React.MouseEvent) => void;
    dragItem?: DraggedLinkContext;
}> = ({ options, defaultClassName, onClick, dragItem }) => {
    if (!dragItem && !onClick) {
        return null;
    }
    const [{ isDragging }, drag] = dragItem
        ? useDrag({
              item: dragItem,
              collect: (monitor) => ({ isDragging: !!monitor.isDragging() })
          })
        : disabledDragging;
    const baseClassName = (options && options.className) || `menu-item ${defaultClassName}`;
    return (
        <span
            className={`${baseClassName}${isDragging ? " dragging" : ""}`}
            {...{ title: options ? options.title : undefined, ref: drag, onClick }}
        />
    );
};
