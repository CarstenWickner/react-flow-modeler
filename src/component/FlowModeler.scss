@import "node_modules/mathsass/dist/math";

.flow-modeler {
    display: grid;
    padding: 2em 1em;
    align-content: stretch;
    justify-content: stretch;

    $strokeColor: #01579b;
    $dropStrokeColor: #00b0ff;
    $strokeStrength: 0.25em;
    $arrowSize: 1.5 * $strokeStrength;
    $strokeLength: 1.75em;
    $elementSizeFactor: 2.5;
    $standardElementSize: $elementSizeFactor * 1em;
    $gatewayBorderSize: sqrt($elementSizeFactor * $elementSizeFactor * 2) * 1em / 2;
    $standardElementColor: #808080;
    $selectionColor: #ffab00;

    & .grid-cell {
        display: flex;
        align-items: center;
        align-content: center;
        min-height: $strokeLength;
        position: relative;
    }
    & .flow-element-wrapper {
        display: flex;
        align-items: center;
        position: relative;
    }

    & .flow-element {
        display: flex;
        position: relative;
        margin: $strokeStrength 0;

        &.start-element {
            border: ($strokeStrength / 2) solid $standardElementColor;
            height: $standardElementSize - $strokeStrength;
            width: $standardElementSize - $strokeStrength;
            border-radius: 50%;
        }
        &.end-element {
            box-shadow: 0 0 0 ($strokeStrength * 0.9) $standardElementColor;
            border: ($strokeStrength * 0.9) solid white;
            background-color: $standardElementColor;
            height: $standardElementSize - (4 * 0.9 * $strokeStrength);
            width: $standardElementSize - (4 * 0.9 * $strokeStrength);
            border-radius: 50%;
            margin: $strokeStrength ($strokeStrength * 0.9);
        }
        &.step-element {
            border: ($strokeStrength / 3) solid $standardElementColor;
            border-radius: 0.25em;
            min-width: $standardElementSize;
            min-height: $standardElementSize / 2;
        }
        &.gateway-element {
            position: relative;
            height: $standardElementSize;
            width: $standardElementSize;
            margin: 0 $strokeStrength / 2;
            z-index: 1;

            display: flex;
            align-items: center;
            justify-content: center;

            &::before {
                content: "";
                position: absolute;
                transform: rotate(45deg);
                border: ($strokeStrength / 2) solid $standardElementColor;
                height: $gatewayBorderSize;
                width: $gatewayBorderSize;
                background-color: transparent;
            }
            &::after {
                content: "X";
                position: absolute;
                font: bold 1.75em Arial, sans-serif;
                color: $standardElementColor;
            }
        }
        &.start-element.selected,
        &.step-element.selected,
        &.gateway-element.selected::before {
            border-color: $selectionColor;
        }
    }
    & .menu {
        position: absolute;
        display: flex;
        flex-direction: column;
        align-items: center;
        left: calc(100% + #{$strokeStrength});
        width: 1em;
        margin: auto;
        border: ($strokeStrength / 3) solid $selectionColor;
        border-radius: 0.25em;
        background-color: white;
        box-shadow: 0.1em 0.1em 0.25em 0.25em rgba(0, 0, 0, 0.25);
        z-index: 1;
        padding: $strokeStrength;

        &-item {
            font-family: Arial, Helvetica, sans-serif;
            line-height: 1.1em;
            cursor: pointer;
    
            &.add-branch::before {
                content: "↳";
            }
            &.add-step::before {
                content: "■";
            }
            &.add-gateway::before {
                content: "◆";
            }
            &.change-next::before {
                content: "⇥";
            }
            &.remove::before {
                content: "×";
            }
        }
    }
    & .gateway-element.diverging + .menu {
        top: ($strokeStrength * 3) + (1.1em / 2);
    }
    & .stroke-horizontal::before,
    & .stroke-vertical::before {
        background-color: $strokeColor;
    }
    &.editable .stroke-horizontal.selected::before {
        background-color: $selectionColor;
    }
    & .stroke-horizontal {
        position: relative;
        display: flex;
        flex-direction: column;
        flex-grow: 1;
        min-width: $strokeLength;

        &.optional {
            min-width: 0;
        }
        &.arrow {
            margin-right: $arrowSize;
    
            &::after {
                content: "";
                height: 0;
                width: 0;
                position: absolute;
                right: 0;
                top: calc(50% - #{$arrowSize});
                margin-right: -$arrowSize;
                border-top: $arrowSize solid transparent;
                border-left: $arrowSize solid $strokeColor;
                border-bottom: $arrowSize solid transparent;
            }
            &.can-drop::after {
                border-left-color: $dropStrokeColor;
            }
        }
        &::before {
            position: absolute;
            content: "";
            height: $strokeStrength;
            width: 100%;
            top: calc(50% - #{$strokeStrength / 2});
        }
        &.can-drop::before {
            background: linear-gradient(90deg, $strokeColor 0%, $dropStrokeColor 62%);
        }
        & > .top-label {
            display: flex;
            flex-direction: column;
            justify-content: flex-end;
            align-content: center;
            padding: 0 0.25em;
            margin-bottom: $strokeStrength / 2;
            min-height: $strokeStrength * 4;
        }
        & > .bottom-spacing {
            display: flex;
            flex-direction: column;
            margin-bottom: $strokeStrength / 2;
        }
        & .clickable-spacing {
            min-height: $strokeStrength * 4;
            max-height: $strokeStrength * 8;
            width: 100%;
        }
    }
    & .stroke-vertical {
        position: relative;
        display: flex;
        flex-direction: column;
        align-self: stretch;
        width: $strokeStrength;
        &::before {
            position: absolute;
            content: "";
            width: $strokeStrength;
        }
        &.top-half::before {
            height: calc(50% + #{$strokeStrength / 2});
            top: 0;
        }
        &.full-height::before {
            height: 100%;
            top: 0;
        }
        &.bottom-half::before {
            height: calc(50% + #{$strokeStrength / 2});
            bottom: 0;
        }
    }
    &.editable {
        & .flow-element,
        & .top-label,
        & .clickable-spacing {
            cursor: context-menu;
        }
        & .flow-element.end-element,
        & .selected.flow-element,
        & .selected .top-label,
        & .selected .clickable-spacing {
            cursor: inherit;
        }
        & .start-element.selected,
        & .start-element:hover,
        & .step-element.selected,
        & .step-element:hover,
        & .gateway-element.selected::before,
        & .gateway-element:hover::before {
            box-shadow: 0 0 0.1em 0.1em rgba(0, 0, 0, 0.25);
        }
        & .flow-element-wrapper.can-drop > .step-element,
        & .flow-element-wrapper.can-drop > .gateway-element::before {
            box-shadow: 0 0 0.125em 0.125em rgba(0, 176, 255, 0.5);
        }
        & .flow-element-wrapper.can-drop > .end-element {
            box-shadow:
                0 0 0 ($strokeStrength * 0.9) $standardElementColor,
                0 0 0.125em ($strokeStrength * 0.9 + 0.125em) rgba(0, 176, 255, 0.5);
        }
        & .dragging {
            color: $dropStrokeColor;
        }
        & .flow-element-wrapper.cannot-drop,
        & .flow-element-wrapper.cannot-drop > .flow-element {
            cursor: no-drop;
        }
    }
}
