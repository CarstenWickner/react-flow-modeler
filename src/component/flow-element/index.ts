export { Start } from "./Start";
export { ContentElement } from "./ContentElement";
export { Gateway } from "./Gateway";
export { GatewayConnector } from "./GatewayConnector";
export { StrokeExtension } from "./StrokeExtension";
export { End } from "./End";
