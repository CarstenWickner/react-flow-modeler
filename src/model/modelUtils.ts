import {
    ContentNode,
    ConvergingGatewayBranch,
    ConvergingGatewayNode,
    DivergingGatewayBranch,
    DivergingGatewayNode,
    ElementType,
    EndNode,
    ModelElement,
    StartNode
} from "./ModelElement";
import { determineRowCounts } from "./rowCountUtils";
import { FlowModelerProps, FlowContent, FlowGatewayDiverging } from "../types/FlowModelerProps";

export const isDivergingGateway = (inputElement: FlowContent | FlowGatewayDiverging): inputElement is FlowGatewayDiverging =>
    !!inputElement && (inputElement as FlowGatewayDiverging).nextElements !== undefined;

export const isMatchingModelElement = (elementId: string): ((element: ModelElement) => element is ContentNode | DivergingGatewayNode | EndNode) => (
    element: ModelElement
): element is ContentNode | DivergingGatewayNode | EndNode =>
    ((element.type === ElementType.Content || element.type === ElementType.GatewayDiverging) && element.id === elementId) ||
    (elementId === null && element.type === ElementType.End);

const createEndNode = (precedingElement: StartNode | ContentNode | ConvergingGatewayNode, resultingModelElements: Array<ModelElement>): EndNode => {
    const end: EndNode = { type: ElementType.End, precedingElement, columnIndex: undefined, rowCount: undefined };
    resultingModelElements.push(end);
    return end;
};

const addAdditionalBranchToConvergingGateway = (
    elementBehindConvergingGateway: ContentNode | DivergingGatewayNode | EndNode,
    resultingModelElements: Array<ModelElement>,
    precedingElement: ContentNode | DivergingGatewayBranch
): ConvergingGatewayBranch => {
    let convergingGateway: ConvergingGatewayNode;
    if (elementBehindConvergingGateway.precedingElement.type === ElementType.GatewayConverging) {
        convergingGateway = elementBehindConvergingGateway.precedingElement;
    } else {
        convergingGateway = {
            type: ElementType.GatewayConverging,
            precedingBranches: [],
            followingElement: elementBehindConvergingGateway,
            columnIndex: undefined,
            rowCount: undefined
        };
        const existingBranch: ConvergingGatewayBranch = {
            type: ElementType.ConnectElementToGateway,
            precedingElement: (elementBehindConvergingGateway.precedingElement as unknown) as ContentNode | DivergingGatewayBranch,
            followingElement: convergingGateway,
            branchIndex: 0,
            columnIndex: undefined,
            rowCount: undefined
        };
        convergingGateway.precedingBranches.push(existingBranch);
        resultingModelElements.push(existingBranch, convergingGateway);
        elementBehindConvergingGateway.precedingElement.followingElement = existingBranch;
        elementBehindConvergingGateway.precedingElement = convergingGateway;
    }
    const newBranch: ConvergingGatewayBranch = {
        type: ElementType.ConnectElementToGateway,
        precedingElement,
        followingElement: elementBehindConvergingGateway.precedingElement,
        branchIndex: convergingGateway.precedingBranches.length,
        columnIndex: undefined,
        rowCount: undefined
    };
    convergingGateway.precedingBranches.push(newBranch);
    resultingModelElements.push(newBranch);
    return newBranch;
};

const handleNextElement = (
    id: string,
    inputElements: FlowModelerProps["flow"]["elements"],
    resultingModelElements: Array<ModelElement>,
    precedingElement: StartNode | ContentNode | DivergingGatewayBranch | ConvergingGatewayNode
): ContentNode | DivergingGatewayNode | ConvergingGatewayBranch | EndNode => {
    if (precedingElement.type !== ElementType.Start) {
        const existingElement = resultingModelElements.find(isMatchingModelElement(id in inputElements ? id : null));
        if (existingElement) {
            return addAdditionalBranchToConvergingGateway(
                existingElement,
                resultingModelElements,
                (precedingElement as unknown) as ContentNode | DivergingGatewayBranch
            );
        }
    }
    const inputElement = inputElements[id];
    if (!inputElement) {
        return createEndNode((precedingElement as unknown) as StartNode | ContentNode | ConvergingGatewayNode, resultingModelElements);
    }
    if (isDivergingGateway(inputElement)) {
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        return createDivergingGatewayNode(id, inputElement, inputElements, resultingModelElements, precedingElement);
    }
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return createContentNode(id, inputElement, inputElements, resultingModelElements, precedingElement);
};

const createContentNode = (
    id: string,
    inputElement: FlowContent,
    inputElements: FlowModelerProps["flow"]["elements"],
    resultingModelElements: Array<ModelElement>,
    precedingElement: StartNode | ContentNode | DivergingGatewayBranch | ConvergingGatewayNode
): ContentNode => {
    const content: ContentNode = {
        type: ElementType.Content,
        id,
        precedingElement,
        followingElement: undefined,
        data: inputElement.data,
        columnIndex: undefined,
        rowCount: undefined
    };
    resultingModelElements.push(content);
    content.followingElement = handleNextElement(inputElement.nextElementId, inputElements, resultingModelElements, content);
    return content;
};

const createDivergingGatewayNode = (
    id: string,
    inputElement: FlowGatewayDiverging,
    inputElements: FlowModelerProps["flow"]["elements"],
    resultingModelElements: Array<ModelElement>,
    precedingElement: StartNode | ContentNode | DivergingGatewayBranch | ConvergingGatewayNode
): DivergingGatewayNode => {
    const divergingGateway: DivergingGatewayNode = {
        type: ElementType.GatewayDiverging,
        id,
        precedingElement,
        followingBranches: [],
        data: inputElement.data,
        columnIndex: undefined,
        rowCount: undefined
    };
    resultingModelElements.push(divergingGateway);
    // ensure that there are always at least two sub elements under a gateway to allow for respective connectors to be displayed
    let subElements: FlowGatewayDiverging["nextElements"];
    if (inputElement.nextElements.length > 1) {
        subElements = inputElement.nextElements;
    } else if (inputElement.nextElements.length === 1) {
        subElements = [...inputElement.nextElements, {}];
    } else {
        subElements = [{}, {}];
    }
    subElements.forEach((branchTarget, branchIndex) => {
        const branch: DivergingGatewayBranch = {
            type: ElementType.ConnectGatewayToElement,
            precedingElement: divergingGateway,
            followingElement: undefined,
            branchIndex,
            data: branchTarget.conditionData,
            columnIndex: undefined,
            rowCount: undefined
        };
        divergingGateway.followingBranches.push(branch);
        resultingModelElements.push(branch);
        // there can be no end node before the next converging gateway
        branch.followingElement = (handleNextElement(branchTarget.id, inputElements, resultingModelElements, branch) as unknown) as
            | ContentNode
            | DivergingGatewayNode
            | ConvergingGatewayBranch;
    });
    return divergingGateway;
};

const determineColumnIndex = (target: ModelElement): number => {
    if (!target.columnIndex) {
        if (target.type === ElementType.GatewayConverging) {
            target.columnIndex = 1 + Math.max(...target.precedingBranches.map(determineColumnIndex));
        } else if (target.type === ElementType.Start) {
            target.columnIndex = 1;
        } else {
            target.columnIndex = 1 + determineColumnIndex(target.precedingElement);
        }
    }
    return target.columnIndex;
};

export const createMinimalElementTreeStructure = ({
    firstElementId,
    elements
}: FlowModelerProps["flow"]): { start: StartNode; elementsInTree: Array<ModelElement> } => {
    const start: StartNode = {
        type: ElementType.Start,
        followingElement: undefined,
        columnIndex: 1,
        rowCount: undefined
    };
    const elementsInTree: Array<ModelElement> = [start];
    const inputElement = elements[firstElementId];
    let firstElement: ContentNode | DivergingGatewayNode | EndNode;
    if (!inputElement) {
        firstElement = createEndNode(start, elementsInTree);
    } else if (isDivergingGateway(inputElement)) {
        firstElement = createDivergingGatewayNode(firstElementId, inputElement, elements, elementsInTree, start);
    } else {
        firstElement = createContentNode(firstElementId, inputElement, elements, elementsInTree, start);
    }
    start.followingElement = firstElement;
    return { start, elementsInTree };
};

export const createElementTree = (flow: FlowModelerProps["flow"], verticalAlign: "top" | "bottom"): StartNode => {
    const { start, elementsInTree } = createMinimalElementTreeStructure(flow);
    // calculate column indexes within the grid to be rendered
    elementsInTree.forEach(determineColumnIndex);
    // calculate number of rows within the grid to be rendered
    determineRowCounts(start, verticalAlign, elementsInTree.forEach.bind(elementsInTree));
    return start;
};
