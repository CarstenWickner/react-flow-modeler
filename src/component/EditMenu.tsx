import * as React from "react";

import { EditMenuItem } from "./EditMenuItem";
import { ContentNode, ConvergingGatewayNode, DivergingGatewayNode, DivergingGatewayBranch, ElementType, StartNode } from "../model/ModelElement";
import { addContentElement, addDivergingGateway } from "../model/action/addElement";
import { addBranch } from "../model/action/addBranch";
import { isChangeNextElementAllowed } from "../model/action/changeNextElement";
import { removeElement, isRemoveElementAllowed } from "../model/action/removeElement";

import { EditActionResult, DraggableType, DraggedLinkContext } from "../types/EditAction";
import { FlowModelerProps } from "../types/FlowModelerProps";

const onClickStopPropagation = (event: React.MouseEvent): void => event.stopPropagation();

export class EditMenu extends React.Component<{
    referenceElement: StartNode | ContentNode | ConvergingGatewayNode | DivergingGatewayNode | DivergingGatewayBranch;
    menuOptions?: FlowModelerProps["options"]["editActions"];
    onChange: (change: (originalFlow: FlowModelerProps["flow"]) => EditActionResult) => void;
}> {
    renderMenuItem(
        options:
            | {
                  className?: string;
                  title?: string;
              }
            | undefined,
        defaultClassName: string,
        onClick?: (event: React.MouseEvent) => void,
        dragType?: DraggableType
    ): React.ReactNode {
        const { referenceElement } = this.props;
        const dragItem: DraggedLinkContext =
            dragType === DraggableType.LINK &&
            (referenceElement.type === ElementType.Content || referenceElement.type === ElementType.ConnectGatewayToElement) &&
            isChangeNextElementAllowed(referenceElement)
                ? { type: dragType, originElement: referenceElement }
                : undefined;
        return <EditMenuItem key={defaultClassName} options={options} defaultClassName={defaultClassName} onClick={onClick} dragItem={dragItem} />;
    }

    onAddContentElementClick = (): void => {
        const { onChange, referenceElement } = this.props;
        if (referenceElement.type !== ElementType.GatewayDiverging) {
            onChange((originalFlow) => addContentElement(originalFlow, referenceElement, {}));
        }
    };

    renderAddContentElementItem(): React.ReactNode {
        const { referenceElement, menuOptions } = this.props;
        if (referenceElement.type === ElementType.GatewayDiverging) {
            return null;
        }
        const options = menuOptions ? menuOptions.addFollowingContentElement : undefined;
        if (options && options.isActionAllowed && !options.isActionAllowed(referenceElement)) {
            return null;
        }
        return this.renderMenuItem(options, "add-content", this.onAddContentElementClick);
    }

    onAddDivergingGatewayClick = (): void => {
        const { referenceElement, onChange } = this.props;
        if (!referenceElement || referenceElement.type !== ElementType.GatewayDiverging) {
            onChange((originalFlow) =>
                addDivergingGateway(
                    originalFlow,
                    (referenceElement as unknown) as ContentNode | ConvergingGatewayNode | DivergingGatewayBranch | undefined,
                    {}
                )
            );
        }
    };

    renderAddDivergingGatewayItem(): React.ReactNode {
        const { referenceElement, menuOptions } = this.props;
        if (referenceElement && referenceElement.type === ElementType.GatewayDiverging) {
            return null;
        }
        const options = menuOptions ? menuOptions.addFollowingDivergingGateway : undefined;
        if (
            options &&
            options.isActionAllowed &&
            !options.isActionAllowed((referenceElement as unknown) as undefined | ContentNode | DivergingGatewayBranch | ConvergingGatewayNode)
        ) {
            return null;
        }
        return this.renderMenuItem(options, "add-gateway", this.onAddDivergingGatewayClick);
    }

    onAddDivergingBranchClick = (): void => {
        const { referenceElement, onChange } = this.props;
        if (referenceElement && referenceElement.type === ElementType.GatewayDiverging) {
            onChange((originalFlow) => addBranch(originalFlow, referenceElement, {}));
        }
    };

    renderAddDivergingBranchItem(): React.ReactNode {
        const { referenceElement, menuOptions } = this.props;
        if (referenceElement.type !== ElementType.GatewayDiverging) {
            return null;
        }
        const options = menuOptions ? menuOptions.addDivergingBranch : undefined;
        if (options && options.isActionAllowed && !options.isActionAllowed(referenceElement)) {
            return null;
        }
        return this.renderMenuItem(options, "add-branch", this.onAddDivergingBranchClick);
    }

    renderChangeNextElementItem(): React.ReactNode {
        const { referenceElement, menuOptions } = this.props;
        if (
            (referenceElement.type !== ElementType.Content && referenceElement.type !== ElementType.ConnectGatewayToElement) ||
            !isChangeNextElementAllowed(referenceElement)
        ) {
            return null;
        }
        const options = menuOptions ? menuOptions.changeNextElement : undefined;
        if (options && options.isActionAllowed && !options.isActionAllowed(referenceElement)) {
            return null;
        }
        return this.renderMenuItem(options, "change-next", undefined, DraggableType.LINK);
    }

    onRemoveClick = (): void => {
        const { referenceElement, onChange } = this.props;
        if (referenceElement.type === ElementType.Content || referenceElement.type === ElementType.ConnectGatewayToElement) {
            onChange((originalFlow) => removeElement(originalFlow, referenceElement));
        }
    };

    renderRemoveItem(): React.ReactNode {
        const { referenceElement, menuOptions } = this.props;
        if (
            (referenceElement.type === ElementType.Content || referenceElement.type === ElementType.ConnectGatewayToElement) &&
            isRemoveElementAllowed(referenceElement)
        ) {
            return this.renderMenuItem(menuOptions ? menuOptions.removeElement : undefined, "remove", this.onRemoveClick);
        }
        return null;
    }

    render(): React.ReactNode {
        const menuItems = [
            this.renderAddContentElementItem(),
            this.renderAddDivergingGatewayItem(),
            this.renderAddDivergingBranchItem(),
            this.renderChangeNextElementItem(),
            this.renderRemoveItem()
        ];
        if (menuItems.some((item) => item)) {
            return (
                <div className="menu" onClick={onClickStopPropagation}>
                    {menuItems}
                </div>
            );
        }
        return null;
    }
}
