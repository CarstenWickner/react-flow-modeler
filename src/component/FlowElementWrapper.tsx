import * as React from "react";
import { useDrop } from "react-dnd";

import { isFlowValid } from "../model/pathValidationUtils";

import { ContentNode, ConvergingGatewayNode, DivergingGatewayNode, ElementType, EndNode, ModelElementExclStart } from "../types/ModelElement";
import { DraggableType, DraggedLinkContext, onLinkDropCallback } from "../types/EditAction";

const isTargetAncestorOfDragItem = (
    originElement: ModelElementExclStart,
    referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode
): boolean => {
    if (originElement === referenceElement) {
        return true;
    }
    if (originElement.type === ElementType.ConvergingGatewayNode) {
        return originElement.precedingBranches.some((branch) => isTargetAncestorOfDragItem(branch, referenceElement));
    }
    return (
        originElement.precedingElement.type !== ElementType.StartNode && isTargetAncestorOfDragItem(originElement.precedingElement, referenceElement)
    );
};

const isDropValid = (referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode, onDrop: onLinkDropCallback) => (
    dragContext: DraggedLinkContext
): boolean =>
    !isTargetAncestorOfDragItem(dragContext.originElement, referenceElement) &&
    dragContext.originElement.followingElement !== referenceElement &&
    isFlowValid(onDrop(referenceElement, dragContext, true).changedFlow);

const disabledDropping: [{ isOver: boolean; canDrop: false }, React.LegacyRef<HTMLDivElement>] = [{ isOver: false, canDrop: false }, undefined];

export const FlowElementWrapper: React.FC<{
    elementTypeClassName: string;
    referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode;
    editMenu?: (() => React.ReactNode) | undefined;
    onLinkDrop?: onLinkDropCallback | undefined;
    onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
}> = ({ elementTypeClassName, referenceElement, editMenu, onLinkDrop, onClick, children }) => {
    const [{ isOver, canDrop }, drop] = onLinkDrop
        ? useDrop({
              accept: DraggableType.LINK,
              canDrop: isDropValid(referenceElement, onLinkDrop),
              drop: (dragContext: DraggedLinkContext) => onLinkDrop(referenceElement, dragContext),
              collect: (monitor) => ({
                  isOver: !!monitor.isOver(),
                  canDrop: !!monitor.canDrop()
              })
          })
        : disabledDropping;
    return (
        <>
            <div className={`stroke-horizontal arrow${isOver && canDrop ? " can-drop" : ""}`} />
            <div className={`flow-element-wrapper${isOver ? (canDrop ? " can-drop" : " cannot-drop") : ""}`} ref={drop}>
                <div className={`flow-element ${elementTypeClassName}${editMenu ? " selected" : ""}`} onClick={onClick}>
                    {children}
                </div>
                {editMenu && editMenu()}
            </div>
        </>
    );
};
