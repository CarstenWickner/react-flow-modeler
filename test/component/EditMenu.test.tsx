import * as React from "react";
import { shallow } from "enzyme";

import { EditMenu } from "../../src/component/EditMenu";
import { createMinimalElementTreeStructure } from "../../src/model/modelUtils";
import {
    StartNode,
    ElementType,
    ModelElement,
    ContentNode,
    ConvergingGatewayNode,
    DivergingGatewayBranch,
    DivergingGatewayNode
} from "../../src/model/ModelElement";

describe("renders correctly", () => {
    const { elementsInTree }: { elementsInTree: Array<ModelElement> } = createMinimalElementTreeStructure({
        firstElementId: "a",
        elements: {
            a: { nextElements: [{ id: "b" }, { id: "c" }, {}] },
            b: { nextElementId: "c" },
            c: {}
        }
    });
    const onChange = (): void => {};
    it("for start element", () => {
        const component = shallow(
            <EditMenu referenceElement={elementsInTree.find((entry) => entry.type === ElementType.Start) as StartNode} onChange={onChange} />
        );
        expect(component).toMatchSnapshot();
    });
    it("for content element", () => {
        const component = shallow(
            <EditMenu
                referenceElement={elementsInTree.find((entry) => entry.type === ElementType.Content && entry.id === "b") as ContentNode}
                onChange={onChange}
            />
        );
        expect(component).toMatchSnapshot();
    });
    it("for diverging gateway", () => {
        const component = shallow(
            <EditMenu
                referenceElement={
                    elementsInTree.find((entry) => entry.type === ElementType.GatewayDiverging && entry.id === "a") as DivergingGatewayNode
                }
                onChange={onChange}
            />
        );
        expect(component).toMatchSnapshot();
    });
    it("for converging gateway", () => {
        const component = shallow(
            <EditMenu
                referenceElement={
                    elementsInTree.find(
                        (entry) =>
                            entry.type === ElementType.GatewayConverging &&
                            entry.followingElement.type === ElementType.Content &&
                            entry.followingElement.id === "c"
                    ) as ConvergingGatewayNode
                }
                onChange={onChange}
            />
        );
        expect(component).toMatchSnapshot();
    });
    it("for diverging gateway branch", () => {
        const component = shallow(
            <EditMenu
                referenceElement={
                    elementsInTree.find(
                        (entry) => entry.type === ElementType.ConnectGatewayToElement && entry.precedingElement.id === "a" && entry.branchIndex === 1
                    ) as DivergingGatewayBranch
                }
                onChange={onChange}
            />
        );
        expect(component).toMatchSnapshot();
    });
});
