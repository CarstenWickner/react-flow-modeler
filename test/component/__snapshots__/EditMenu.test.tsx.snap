// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly for content element 1`] = `
<div
  className="menu"
  onClick={[Function]}
>
  <Component
    defaultClassName="add-content"
    key="add-content"
    onClick={[Function]}
  />
  <Component
    defaultClassName="add-gateway"
    key="add-gateway"
    onClick={[Function]}
  />
  <Component
    defaultClassName="change-next"
    dragItem={
      Object {
        "originElement": ContentNode { "id": "b" },
        "type": "link",
      }
    }
    key="change-next"
  />
  <Component
    defaultClassName="remove"
    key="remove"
    onClick={[Function]}
  />
</div>
`;

exports[`renders correctly for converging gateway 1`] = `
<div
  className="menu"
  onClick={[Function]}
>
  <Component
    defaultClassName="add-content"
    key="add-content"
    onClick={[Function]}
  />
  <Component
    defaultClassName="add-gateway"
    key="add-gateway"
    onClick={[Function]}
  />
</div>
`;

exports[`renders correctly for diverging gateway 1`] = `
<div
  className="menu"
  onClick={[Function]}
>
  <Component
    defaultClassName="add-branch"
    key="add-branch"
    onClick={[Function]}
  />
</div>
`;

exports[`renders correctly for diverging gateway branch 1`] = `
<div
  className="menu"
  onClick={[Function]}
>
  <Component
    defaultClassName="add-content"
    key="add-content"
    onClick={[Function]}
  />
  <Component
    defaultClassName="add-gateway"
    key="add-gateway"
    onClick={[Function]}
  />
  <Component
    defaultClassName="change-next"
    dragItem={
      Object {
        "originElement": DivergingGatewayBranch { "followingElement": ConvergingGatewayBranch { "gateway": ConvergingGatewayNode { "followingElement": ContentNode { "id": "c" } } } },
        "type": "link",
      }
    }
    key="change-next"
  />
  <Component
    defaultClassName="remove"
    key="remove"
    onClick={[Function]}
  />
</div>
`;

exports[`renders correctly for start element 1`] = `
<div
  className="menu"
  onClick={[Function]}
>
  <Component
    defaultClassName="add-content"
    key="add-content"
    onClick={[Function]}
  />
  <Component
    defaultClassName="add-gateway"
    key="add-gateway"
    onClick={[Function]}
  />
</div>
`;
